### ISSUE TITLE
isort is really slow when using --recursive since no way to efficiently exclude directories

### ISSUE THREAD

**Issue by edmorley** [timothycrosley/isort#337]
I've run isort against a repo and made some corrections, however to prevent regressions, I need to:
(a) Run on travis
(b) Make it quick for people to run locally before pushing, so they don't keep on getting burnt jobs on travis

To do this, I'd like to use:
`isort --recursive --check-only --diff .`

However this takes a long time to run (at least locally from inside the virtualenv in my vagrant environment, where the source is mounted using virtualbox shared folders).

ie: 36 seconds, compared to 5 seconds for running flake8 recursively against the same repo.

Annotating the isort file shows that it's spending a lot of time iterating over the `.git/` and `node_modules` directories, even though I have them listed in `--skip`.

Inspecting the source shows that isort doesn't use the contents of `--skip` (or `--skip-glob`) when it's constructing the list of files to scan. Instead it waits until after it's instantiated the `SortImports` instance to use them.

Compare this to pep8's implementation, where it skips a whole subdirectory if appropriate:
https://github.com/PyCQA/pep8/blob/b1bde9f2bbcad0f3c86960e58c7cb98ec0cbfd77/pep8.py#L1827-L1888

As a workaround, I'm having to hardcode the list of subdirectories, eg:
`isort --recursive --check-only --verbose *.py puppet/ tests/ treeherder/`
...which runs in 2.5 seconds.

Compared to the original, which takes 36 seconds - but is scanning the same set of Python files!

Please could a way be added to skip whole subdirectories in a performant manner? :-)

**Comment by timothycrosley** [comment]
Thanks for reporting this major performance issues! One of my primary goals with isort is for it to run quickly over entire projects, so that individuals can use it without hesitation. I'll make sure to resolve this before the next release.

Thanks!

~Timothy

**Comment by edmorley** [comment]
Many thanks! :-)

**Comment by timothycrosley** [comment]
This should be fixed in the just released 4.2.0,

Thanks!

~Timothy

**Comment by edmorley** [comment]
```bash
vagrant ~/treeherder $ time isort --check-only
<snip>

real    0m2.116s
user    0m0.772s
sys     0m0.332s
```

Amazing - and thank you for the really quick turnaround :-)
